cmd/metal-api/internal/fsm/states/waiting.go

```diff
@@ -2,21 +2,28 @@ package states
 
 import (
 	"github.com/looplab/fsm"
-	"github.com/metal-stack/metal-api/cmd/metal-api/internal/metal"
 )
 
 type WaitingState struct {
-	container *metal.ProvisioningEventContainer
-	event     *metal.ProvisioningEvent
+	config *StateConfig
 }
 
 func newWaiting(c *StateConfig) *WaitingState {
 	return &WaitingState{
-		container: c.Container,
-		event:     c.Event,
+		config: c,
 	}
 }
 
 func (p *WaitingState) OnTransition(e *fsm.Event) {
-	appendEventToContainer(p.event, p.container)
+	appendEventToContainer(p.config.Event, p.config.Container)
+
+	if p.config.Scaler != nil {
+		e.Err = p.config.Scaler.AdjustNumberOfWaitingMachines(p.config.Partition)
+	}
```

Comment:
Can you explain to me why we need to do this `OnTransition` and then again on `OnLeave`? Why not only `OnLeave`?

Reply:
`OnTransition` happens when waiting state is entered, which should also be handled as far as I understand. Maybe we should call this `OnEnter` instead.

Reply:
In the transition we append the event to the container, but why do we need to call the scaler? Isn't the current implementation calling the scaler two times in a row in a matter of microseconds? In this short period of time, the state in the database will not have changed?

Reply:
If a machine changes from some previous state to waiting, the fsm will call OnTransition but not OnLeave.

Reply:
Okay, sorry, I got it now. I think the `OnTransition` triggered the wrong thoughts in my brain. Maybe `OnEnter` would really be better.